### PR TITLE
contrib: Add ContainerLab-based BGP CPlane development environment

### PR DESCRIPTION
**Pull request by YutaroHayakawa**
Please see the commit message for more details.

```release-note
contrib: Add ContainerLab-based BGP CPlane development environment
```

### CODE REVIEW

**Review by rastislavs — APPROVED** [review]
nice!

**Comment by YutaroHayakawa** [comment]
@aditighag @jibi @learnitall  Please kindly take a look. It shouldn't take time since it doesn't interfere with any existing things.

**Review by jibi — APPROVED** [review]

File: Documentation/contributing/development/bgp_cplane.rst [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+=================
+BGP Control Plane
+=================
+
+This section is specific to BGP Control Plane contributions.
+
+Development Environment
+=======================
+
+BGP Control Plane requires BGP peer for testing. You can use `ContainerLab <https://containerlab.dev/>`_ and `Kind <https://kind.sigs.k8s.io/docs/user/configuration/>`_-based development environment. The following diagram shows the topology.
```
Comment (learnitall) [inline Documentation/contributing/development/bgp_cplane.rst]: Couple of suggestions here for grammar and word choice.
```suggestion
BGP Control Plane requires a BGP peer for testing. This section describes a `ContainerLab <https://containerlab.dev/>`_ and `Kind <https://kind.sigs.k8s.io/docs/user/configuration/>`_-based development environment. The following diagram shows the topology:
```

File: Documentation/contributing/development/bgp_cplane.rst [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+.. image:: _static/bgp-cplane-dev-topo.png
+    :align: center
+
+Role of each nodes are following
```
Comment (learnitall) [inline Documentation/contributing/development/bgp_cplane.rst]: I'm not sure if the grammar is right here, I left a suggestion for rewording this.

```suggestion
The following describes the role of each node:
```

File: Documentation/contributing/development/bgp_cplane.rst [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+* ``router0`` is an `FRRouting (FRR) <https://frrouting.org/>`_ router. It is pre-configured with a minimal peering settings with server0 and server1.
```
Comment (learnitall) [inline Documentation/contributing/development/bgp_cplane.rst]: The "a" tells the reader that the noun it applies to is singular, but "settings" is plural. I'd suggest the following:

```suggestion
* ``router0`` is an `FRRouting (FRR) <https://frrouting.org/>`_ router. It is pre-configured with minimal peering settings with server0 and server1.
```

File: Documentation/contributing/development/bgp_cplane.rst [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+* ``server0`` and ``server1`` are Kind nodes (technically, they are ``nicolaka/netshoot`` containers that share network namespaces with Kind nodes).
```
Comment (learnitall) [inline Documentation/contributing/development/bgp_cplane.rst]: Do you think you could reword this so there aren't parentheses used here? This is really cool that these are netshoot containers that share a netns with kind nodes, it shows off the power and usefulness of containerlab. I think it makes sense to highlight that. Here's an example of what I'm thinking:

```suggestion
* ``server0`` and ``server1`` are ``nicolaka/netshoot`` containers that each share a network namespace with their own Kind node.
```

File: Documentation/contributing/development/bgp_cplane.rst [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+   $ make kind-bgp-v4
+
+.. note::
+        Above example setups IPv4 single-stack environment. You can change ``v4`` part to ``v6`` or ``dual`` to setup IPv6 single-stack or dual-stack environment respectively. Same for all below examples.
+
+Install Cilium on the Lab
+-------------------------
+
+Install Cilium on the lab with your favorite way. Following example assumes you are modifying the source and want to build your image. The minimal mandatory Helm values are provided in ``contrib/containerlab/bgp-cplane-dev-v4/values.yaml``. You can add Helm values to deploy BGP Control Plane with different Cilium configuration.
```
Comment (learnitall) [inline Documentation/contributing/development/bgp_cplane.rst]: There are a couple of changes that I'd recommend here for grammar and word choice:

```suggestion
Install Cilium on the lab with your favorite way. The following example assumes you are modifying the source and want to build your own image. The minimal mandatory Helm values are provided in ``contrib/containerlab/bgp-cplane-dev-v4/values.yaml``. If needed, you can add Helm values to deploy BGP Control Plane with a different Cilium configuration.
```

File: Documentation/contributing/development/bgp_cplane.rst (hunk 1 above)
Comment (learnitall) [reply Documentation/contributing/development/bgp_cplane.rst]: Additionally, can you modify these links so they are block-level hyperlink targets? See https://docs.cilium.io/en/stable/contributing/docs/docsstyle/#links

File: Documentation/contributing/development/bgp_cplane.rst [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+.. code-block:: shell-session
+
+   $ make docker-cilium-image docker-operator-generic-image
+   $ kind --name bgp-cplane-dev-v4 load docker-image quay.io/cilium/cilium:latest quay.io/cilium/operator-generic:latest
+   $ cilium install --chart-directory install/kubernetes/cilium -f contrib/containerlab/bgp-cplane-dev-v4/values.yaml --helm-set image.override="quay.io/cilium/cilium:latest" --helm-set image.pullPolicy=Never --helm-set operator.image.override="quay.io/cilium/operator-generic:latest" --helm-set operator.image.pullPolicy=Never
+
+Peering with Router
+-------------------
+
+Peering Cilium nodes with FRR by applying CiliumBGPPeeringPolicy.
```
Comment (learnitall) [inline Documentation/contributing/development/bgp_cplane.rst]: A couple of grammar suggestions to reword this as an instruction for the user:

```suggestion
Peer Cilium nodes with FRR by applying a CiliumBGPPeeringPolicy:
```

File: Documentation/contributing/development/bgp_cplane.rst [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+   $ cilium bgp peers
+
+Validating Connectivity to Cilium Nodes from Non-Cilium Node
+------------------------------------------------------------
+
+.. code-block:: shell-session
+
+   $ docker exec -it clab-bgp-cplane-dev-v4-server2 mtr 10.0.1.2
+   $ docker exec -it clab-bgp-cplane-dev-v4-server2 mtr 10.0.2.2
```
Comment (learnitall) [inline Documentation/contributing/development/bgp_cplane.rst]: Is there any output here that the user should expect if the connectivity is successful? Or is there no output and a zero RC?

**Review by learnitall — CHANGES_REQUESTED** [review]
Thank you for your patience. This is a really cool doc! I just have a couple of changes that I'd recommend for grammar and structure and then we can get this merged in.

**Review by aditighag — APPROVED** [review]

File: Documentation/contributing/development/bgp_cplane.rst (hunk 7 above)
Comment (YutaroHayakawa) [reply Documentation/contributing/development/bgp_cplane.rst]: Good point. Updated with the expected output and description for it.

**Comment by YutaroHayakawa** [comment]
@learnitall Thanks for your review! I applied your suggestion.

I also changed the diagram to show more detailed addressing information.

**Comment by YutaroHayakawa** [comment]
cc: @danehans

File: contrib/containerlab/bgp-cplane-dev-dual/.gitignore [hunk 8]
```diff
@@ -0,0 +1,2 @@
+clab-bgp-cplane-dev-dual/*
+.topo.yaml.bak
```
Comment (danehans) [inline contrib/containerlab/bgp-cplane-dev-dual/.gitignore:4]: I don't see `.topo.yaml.bak` used anywhere other than being referenced by the gitignores. Is this an accidental leftover from testing?

Why ignore all files in this directory?

File: contrib/containerlab/bgp-cplane-dev-dual/topo.yaml [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      - ip route add fd00::/16 via fd00:10:0:1::1 dev net0
+    # Server with Cilium. It shares netns with kind node.
+    server1:
+      kind: linux
+      image: nicolaka/netshoot:v0.11
+      network-mode: container:bgp-cplane-dev-dual-worker
+      exec:
+      - ip addr add 10.0.2.2/24 dev net0
+      - ip addr add fd00:10:0:2::2/64 dev net0
+      - ip route add 10.0.0.0/8 via 10.0.2.1 dev net0
```
Comment (danehans) [inline contrib/containerlab/bgp-cplane-dev-dual/topo.yaml:61]: L49 applies here too, correct?

File: Documentation/contributing/development/bgp_cplane.rst [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    https://docs.cilium.io
+
+
+.. _bgp_cplane_contributing:
+
+=================
+BGP Control Plane
+=================
+
+This section is specific to BGP Control Plane contributions.
```
Comment (danehans) [inline Documentation/contributing/development/bgp_cplane.rst]: nit: consider linking `BGP Control Plane` to the BGP CP [docs](https://docs.cilium.io/en/stable/network/bgp-control-plane/).

File: contrib/containerlab/bgp-cplane-dev-dual/Makefile [hunk 11]
```diff
@@ -0,0 +1,10 @@
+deploy:
+	kind create cluster --config cluster.yaml
+	sudo containerlab -t topo.yaml deploy
```
Comment (danehans) [inline contrib/containerlab/bgp-cplane-dev-dual/Makefile:3]: I already had the `containerlab` bin installed so this target worked fine for me. Should a note be added to the doc that states this requirement and the need for sudo privs? If so, you may also want to include a specific version of the bin.

File: Documentation/contributing/development/bgp_cplane.rst [hunk 12]
```diff
@@ -0,0 +1,99 @@
+.. only:: not (epub or latex or html)
+
```
Comment (danehans) [inline Documentation/contributing/development/bgp_cplane.rst:2]: Adding a comment here since I can't comment on the png file. Consider adding the Kind node name to the diagram, e.g. `server0 (Kind control-plane Node, Cilium)`.

File: Documentation/contributing/development/bgp_cplane.rst [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+.. note::
+        At this point, there are only minimal peering settings on the policy and no advertisement configuration present. You need to edit policies, for example, with ``kubectl edit bgpp`` to realize your desired settings. If you need to change the router side, you can edit FRRouting settings with ``docker exec -it clab-bgp-cplane-dev-v4-router0 vtysh``.
+
+Validating Peering Status
+-------------------------
+
+.. code-block:: shell-session
+
+   $ cilium bgp peers
```
Comment (danehans) [inline Documentation/contributing/development/bgp_cplane.rst:82]: Consider adding the expected response from the command or add a short note that states the expected session state, routers received/advertised, etc.

File: Documentation/contributing/development/bgp_cplane.rst [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+Install Cilium on the Lab
+-------------------------
+
+Install Cilium on the lab with your favorite way. The following example assumes you are modifying the source and want to build your own image. The minimal mandatory Helm values are provided in ``contrib/containerlab/bgp-cplane-dev-v4/values.yaml``. If needed, you can add Helm values to deploy BGP Control Plane with a different Cilium configuration.
+
+.. code-block:: shell-session
+
+   $ make docker-cilium-image docker-operator-generic-image
+   $ kind --name bgp-cplane-dev-v4 load docker-image quay.io/cilium/cilium:latest quay.io/cilium/operator-generic:latest
+   $ cilium install --chart-directory install/kubernetes/cilium -f contrib/containerlab/bgp-cplane-dev-v4/values.yaml --helm-set image.override="quay.io/cilium/cilium:latest" --helm-set image.pullPolicy=Never --helm-set operator.image.override="quay.io/cilium/operator-generic:latest" --helm-set operator.image.pullPolicy=Never
```
Comment (danehans) [inline Documentation/contributing/development/bgp_cplane.rst]: This workflow could be simplified by adding the ability to specify the kind cluster name. For example:

```sh
$ git diff Makefile
diff --git a/Makefile b/Makefile
index 3730ee6c01..bd0d692cb0 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,8 @@
 # Copyright Authors of Cilium
 # SPDX-License-Identifier: Apache-2.0

+CLUSTER_NAME ?= kind
+
 ##@ Default
 all: precheck build postcheck ## Default make target that perform precheck -> build -> postcheck
        @echo "Build finished."
@@ -576,7 +578,7 @@ kind-build-image-agent: ## Build cilium-dev docker image

 $(eval $(call KIND_ENV,kind-image-agent))
 kind-image-agent: kind-ready kind-build-image-agent ## Build cilium-dev docker image and import it into kind.
-       $(QUIET)kind load docker-image $(LOCAL_AGENT_IMAGE)
+       $(QUIET)kind load docker-image $(LOCAL_AGENT_IMAGE) -n $(CLUSTER_NAME)

 $(eval $(call KIND_ENV,kind-build-image-operator))
 kind-build-image-operator: ## Build cilium-operator-dev docker image
@@ -584,7 +586,7 @@ kind-build-image-operator: ## Build cilium-operator-dev docker image

 $(eval $(call KIND_ENV,kind-image-operator))
 kind-image-operator: kind-ready kind-build-image-operator ## Build cilium-operator-dev docker image and import it into kind.
-       $(QUIET)kind load docker-image $(LOCAL_OPERATOR_IMAGE)
+       $(QUIET)kind load docker-image $(LOCAL_OPERATOR_IMAGE) -n $(CLUSTER_NAME)

 $(eval $(call KIND_ENV,kind-build-clustermesh-apiserver))
 kind-build-clustermesh-apiserver: ## Build cilium-clustermesh-apiserver docker image
```

A user would then:

```sh
$ CLUSTER_NAME=bgp-cplane-dev-v4 make kind-image
$ cilium install --chart-directory install/kubernetes/cilium -f contrib/containerlab/bgp-cplane-dev-v4/values.yaml
```

**Review by danehans — COMMENTED** [review]
I took this PR for a test drive and this will be a nice way to test/dev/demo BGP CP.

File: Documentation/contributing/development/bgp_cplane.rst [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+Deploy Lab
+----------
+
+.. code-block:: shell-session
+
+   $ make kind-bgp-v4
+
+.. note::
+        Above example setups IPv4 single-stack environment. You can change ``v4`` part to ``v6`` or ``dual`` to setup IPv6 single-stack or dual-stack environment respectively. Same for all below examples.
```
Comment (learnitall) [inline Documentation/contributing/development/bgp_cplane.rst]: Can you replace the use of "Above" and "below" here? Sometimes translators have a hard time with these words in this context. Using "prior" and "following" are great alternatives.

Additionally, because "setup" is used as a verb here, it needs to be separated into two words. It's something that I had to learn when reviewing this :). This page is really helpful: https://www.grammarbook.com/blog/spelling/setup-vs-set-up/.

I also left some suggestions for grammar.

```suggestion
        The prior example sets up an IPv4 single-stack environment. You can change the ``v4`` part to ``v6`` or ``dual`` to set up an IPv6 single-stack or dual-stack environment respectively. The same goes for the following examples.
```

File: Documentation/contributing/development/bgp_cplane.rst [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+-------------------------
+
+.. code-block:: shell-session
+
+   $ cilium bgp peers
+
+Validating Connectivity to Cilium Nodes from Non-Cilium Node
+------------------------------------------------------------
+
+The below example validates connectivity from server2 to server0 (10.0.1.2) and server1 (10.0.2.2). You should see the packet goes through router0 (10.0.3.1).
```
Comment (learnitall) [inline Documentation/contributing/development/bgp_cplane.rst]: I think this needs to be "go" here instead of "goes"

```suggestion
The below example validates connectivity from server2 to server0 (10.0.1.2) and server1 (10.0.2.2). You should see the packet go through router0 (10.0.3.1).
```

File: Documentation/contributing/development/bgp_cplane.rst [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+.. code-block:: shell-session
+
+   $ docker exec -it clab-bgp-cplane-dev-v4-server2 mtr 10.0.1.2 -r
+   Start: 2023-10-04T02:24:58+0000
+   HOST: server2                     Loss%   Snt   Last   Avg  Best  Wrst StDev
+   1.|-- 10.0.3.1                   0.0%    10    0.0   0.1   0.0   0.1   0.0
```
Comment (learnitall) [inline Documentation/contributing/development/bgp_cplane.rst]: I'm not familiar with the expected output here, but is there some whitespace misalignment here? The data in the columns don't seem to line up with their associated headers. For instance, I would have expected:

```
$ docker exec -it clab-bgp-cplane-dev-v4-server2 mtr 10.0.2.2 -r
Start: 2023-10-04T02:25:58+0000
HOST: server2                     Loss%   Snt   Last   Avg  Best  Wrst StDev
1.|-- 10.0.3.1                     0.0%    10    0.1   0.1   0.0   0.1   0.0
2.|-- 10.0.2.2                     0.0%    10    0.1   0.1   0.0   0.2   0.0
```

The same goes for the table in the second command.

Sometimes weird stuff can happen when copying and pasting command output, so I just wanted to double check.

File: Documentation/contributing/development/bgp_cplane.rst (hunk 13 above)
Comment (learnitall) [reply Documentation/contributing/development/bgp_cplane.rst:73]: +1 from me, that's a great idea

**Review by learnitall — CHANGES_REQUESTED** [review]
The changes look great! I have a couple last notes for you in regards to some of the stuff you changed, as well as something that I accidentally missed in my last review.

File: Documentation/contributing/development/bgp_cplane.rst (hunk 13 above)
Comment (learnitall) [reply Documentation/contributing/development/bgp_cplane.rst:82]: +1 from me, that's a great idea.

File: contrib/containerlab/bgp-cplane-dev-dual/.gitignore (hunk 8 above)
Comment (YutaroHayakawa) [reply contrib/containerlab/bgp-cplane-dev-dual/.gitignore:4]: These are the file/dir automatically generated by ContainerLab. `clab-bgp-cplane-dev-dual` is a directory to store the runtime state and .topo.yaml.bak is a backup manifest generated when you create the lab.

File: contrib/containerlab/bgp-cplane-dev-dual/topo.yaml (hunk 9 above)
Comment (YutaroHayakawa) [reply contrib/containerlab/bgp-cplane-dev-dual/topo.yaml:61]: Correct. Do you mean we should have a comment here as well? I omitted it because it is already mentioned on the doc that the roles of server0 and server1 are the same, and assumed that people could understand what does mean from server0's configuration.

File: contrib/containerlab/bgp-cplane-dev-dual/.gitignore (hunk 8 above)
Comment (danehans) [reply contrib/containerlab/bgp-cplane-dev-dual/.gitignore:4]: Thanks for the input. Might not be a bad idea to document ^ somewhere.

File: contrib/containerlab/bgp-cplane-dev-dual/topo.yaml (hunk 9 above)
Comment (danehans) [reply contrib/containerlab/bgp-cplane-dev-dual/topo.yaml:61]: I don't think adding the same comment throughout is necessary but I don't think it hurts either... your call.

File: Documentation/contributing/development/bgp_cplane.rst (hunk 12 above)
Comment (YutaroHayakawa) [reply Documentation/contributing/development/bgp_cplane.rst:2]: Good idea! I added the control-plane and worker, respectively.

File: Documentation/contributing/development/bgp_cplane.rst (hunk 15 above)
Comment (YutaroHayakawa) [reply Documentation/contributing/development/bgp_cplane.rst]: Thanks for the input! I didn't know they were different words 😮 

Applied your suggestion.

File: Documentation/contributing/development/bgp_cplane.rst (hunk 13 above)
Comment (YutaroHayakawa) [reply Documentation/contributing/development/bgp_cplane.rst:82]: Thanks, I added an expected result and description.

File: Documentation/contributing/development/bgp_cplane.rst (hunk 16 above)
Comment (YutaroHayakawa) [reply Documentation/contributing/development/bgp_cplane.rst]: The `packet` needed to be `packets` as well. Fixed.

File: contrib/containerlab/bgp-cplane-dev-dual/.gitignore (hunk 8 above)
Comment (YutaroHayakawa) [reply contrib/containerlab/bgp-cplane-dev-dual/.gitignore:4]: Good idea, documented in the comments on the `.gitignore` files.

File: contrib/containerlab/bgp-cplane-dev-dual/Makefile (hunk 11 above)
Comment (YutaroHayakawa) [reply contrib/containerlab/bgp-cplane-dev-dual/Makefile:3]: Indeed. I added a new `Prerequisites` section.

File: contrib/containerlab/bgp-cplane-dev-dual/topo.yaml (hunk 9 above)
Comment (YutaroHayakawa) [reply contrib/containerlab/bgp-cplane-dev-dual/topo.yaml:61]: Yep, nothing hurts. I added comments.

File: Documentation/contributing/development/bgp_cplane.rst (hunk 14 above)
Comment (YutaroHayakawa) [reply Documentation/contributing/development/bgp_cplane.rst]: Following the workflow of the `make kind-xxx` target sounds good to me. I put your suggestion into a separate commit and updated the document. Please take a look.

File: Documentation/contributing/development/bgp_cplane.rst (hunk 17 above)
Comment (YutaroHayakawa) [reply Documentation/contributing/development/bgp_cplane.rst]: Yes, as you said, something wrong happened when I copied and pasted the output. I fixed it.

File: Documentation/contributing/development/bgp_cplane.rst (hunk 10 above)
Comment (YutaroHayakawa) [reply Documentation/contributing/development/bgp_cplane.rst]: Good idea. Done.

**Review by danehans — APPROVED** [review]

**Review by learnitall — APPROVED** [review]
Looks great!

**Comment by YutaroHayakawa** [comment]
Conformance K8s Kind: Looks like a new flake https://github.com/cilium/cilium/issues/28475

**Comment by YutaroHayakawa** [comment]
Cilium IPsec upgrade: Known flake https://github.com/cilium/cilium/issues/28088
